src/server/server/dal.py

```diff
@@ -307,5 +47,352 @@ def _to_db_action_type_on_doc(user_action_on_doc_enum):
+            db_profile.key = user._user_computed_profile_db_key  # pylint: disable=protected-access
+            db_profiles.append(db_profile)
+        self.ds_client.put_multi(db_profiles)
+
+    def get_user_computed_profiles(self, users):
+        keys = [user._user_computed_profile_db_key for user in users]  # pylint: disable=protected-access
+        db_profiles = self.ds_client.get_multi(keys)
+        profiles = [_to_user_computed_profile(db_profile) for db_profile in db_profiles]
+        return profiles
+
+    def get_user_feature_vector(self, user):
+        return self.get_users_feature_vectors([user])[0]
+
+    def get_users_feature_vectors(self, users):
+        # NB: this could be probably optimized by projection query if we need, but it would require to index vector property.
+        profiles = self.get_user_computed_profiles(users)
+        return [profile.feature_vector for profile in profiles]
+
+    def _to_user_docs(self, db_user_doc_set):
+        if 'user_documents' not in db_user_doc_set:  # cannot save an empty list in datastore: field would be removed
+            return []
+        db_user_docs = db_user_doc_set['user_documents']
+        db_doc_keys = [user_doc['document_key'] for user_doc in db_user_docs]
+        docs = self._to_docs(db_doc_keys)
+        user_docs = [_to_user_doc(doc, db_user_doc) for doc, db_user_doc in zip(docs, db_user_docs)]
+        return user_docs
+
+    def _to_docs(self, db_doc_keys):
+        db_docs = self.ds_client.get_multi(db_doc_keys)
+        docs = [_to_doc(db_doc) for db_doc in db_docs]
+        return docs
+
+    def save_user_docs(self, user, user_docs):
+        user_doc_set_db_key = user._user_doc_set_db_key  # pylint: disable=protected-access
+        db_user_doc_set = self.ds_client.get(user_doc_set_db_key)
+        db_user_docs = self._to_db_user_docs(user_docs)
+        db_user_doc_set['user_documents'] = db_user_docs
+        self.ds_client.put(db_user_doc_set)
+
+    def _to_db_user_docs(self, user_docs):
+        db_user_docs = [self._to_db_user_doc(user_doc) for user_doc in user_docs]
+        return db_user_docs
+
+    def _to_db_user_doc(self, user_doc):
+        db_user_doc = self._make_entity(u'UserDocument', not_indexed=())
+        db_user_doc['document_key'] = user_doc.document._db_key  # pylint: disable=protected-access
+        db_user_doc['grade'] = user_doc.grade
+        return db_user_doc
+
+    def save_users_docs(self, user_to_user_docs_list):
+        """
+        Save UserDocuments list associated to each User
+        :param user_to_user_docs_list: list of tuples (User, List of UserDocument)
+        """
+        user_set_db_keys = []
+        users_docs = []
+        for user, user_docs in user_to_user_docs_list:
+            user_set_db_keys.append(user._user_doc_set_db_key)  # pylint: disable=protected-access
+            users_docs.append(user_docs)
+        db_user_sets = self.ds_client.get_multi(user_set_db_keys)
```

Comment:
it's not clearly specified in the doc. It is more or less tested though (order of test save_then_get in tests is different). Also the test case here suppose it's in the same order:
https://github.com/GoogleCloudPlatform/gcloud-python/blob/master/gcloud/datastore/test_client.py#L452